File: DFS/77. Combinations/Solution.java
// https://leetcode.com/problems/combinations/

/**
By jiapeng
Runtime: 1 ms, faster than 100.00% of Java online submissions for Combinations.
Memory Usage: 40.8 MB, less than 76.27% of Java online submissions for Combinations.
**/

class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> ans = new ArrayList();
        if (k > n)
            return ans;
        dfs(0, k, n, new ArrayList(), ans, 1);
        return ans;
    }
    
    private void dfs(int d, int k, int n, List<Integer>cur, List<List<Integer>>ans, int starter){
        if(d == k){
            ans.add(new ArrayList(cur));
            return;
        }
        for(int i = starter; i < n+1; i++){
            if(n+1-i < k-d)
                break;
            cur.add(i);
            dfs(d+1, k, n, cur, ans, i+1);
            cur.remove(cur.size()-1);
        }
    }
}

File: DFS/77. Combinations/Solution.py
# https://leetcode.com/problems/combinations/

# By jiapeng
# Runtime: 92 ms, faster than 92.19% of Python3 online submissions for Combinations.
# Memory Usage: 15.2 MB, less than 67.51% of Python3 online submissions for Combinations.

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        if k > n:
            return []
        ans = []
        self.dfs(0, k, n, [], ans, 1)
        return ans
        
    def dfs(self, d, k, n, cur, ans, starter):
        if d == k:
            ans.append(cur)
            return
        for i in range(starter, n+1):
            if n+1-i < k-d:    # if there are less selectable numbers remaining than required, break the loop
                break
            self.dfs(d+1, k, n, cur+[i], ans, i+1)
